src/v0.8/core/collateral/base/escrow.sol

```diff
@@ -0,0 +1,182 @@
+
+    mapping(CollateralType.Type => mapping(address => mapping(uint256 => CollateralType.Collateral)))
+        private datacapCollteralFunds; // mapping(type, mapping(payee, mapping(id, collateral)))
+
+    IDatasets public datasets;
+    /// @dev This empty reserved space is put in place to allow future versions to add new
+    uint256[32] private __gap;
+
+    /// @notice initialize function to initialize the contract and grant the default admin role to the deployer.
+    function initialize(address _roles, address _datasets) public initializer {
+        datasets = IDatasets(_datasets);
+        RolesModifiers.rolesModifiersInitialize(_roles);
+        __Ownable_init();
+        __UUPSUpgradeable_init();
+    }
+
+    /// @notice UUPS Upgradeable function to update the roles implementation
+    /// @dev Only triggered by default admin role
+    function _authorizeUpgrade(
+        address newImplementation
+    )
+        internal
+        override
+        onlyRole(RolesType.DEFAULT_ADMIN_ROLE) // solhint-disable-next-line
+    {}
+
+    /// @notice Returns the implementation contract
+    function getImplementation() external view returns (address) {
+        return _getImplementation();
+    }
+
+    /// @notice transferOwnershipTo newOwner
+    function transferOwnershipTo(address newOwner) public onlyOwner {
+        transferOwnership(newOwner);
+    }
+
+    /// @dev Records the sent amount as credit for future withdrawals.
+    /// @param _payee The destination address for the credited funds.
+    /// @param _id The datasetID or chunkID associated with the credited funds.
+    /// @param _type The collateral type for the credited funds.
+    /// @notice Emits a {Deposited} event upon successful credit recording.
+    function deposit(
+        address _payee,
+        uint64 _id,
+        CollateralType.Type _type
+    ) external payable onlyOwner {
+        uint256 amount = msg.value;
+
+        datacapCollteralFunds[_type][_payee][_id].funds += amount;
+        datacapCollteralFunds[_type][_payee][_id].startedBlockNumber = block
+            .number;
+
+        emit CollateralEvents.Deposited(_payee, amount, _type, _id);
+    }
+
+    /// @dev Retrieves the funds associated with the given payee, datasetID, and collateral type.
+    /// @param _payee The destination address for the funds.
+    /// @param _id The datasetID or chunkID associated with the deposit.
+    /// @param _type The collateral type for the deposit.
+    function depositsOf(
+        address _payee,
+        uint64 _id,
+        CollateralType.Type _type
+    ) external view returns (uint256) {
+        return datacapCollteralFunds[_type][_payee][_id].funds;
+    }
+
+    /// @dev Retrieves the address authorized to withdraw funds.
+    /// @param _payee The destination address for the funds.
+    /// @param _id The datasetID or chunkID associated with the withdrawal.
+    /// @param _type The collateral type for the withdrawal.
+    function withdrawalAllowed(
```

Comment:
can consider using function pointer callback functions for processing

Reply:
add setDatasetsAddrss function to resolve circular dependency